anndata/tests/test_backed_sparse.py

```diff
@@ -212,3 +212,39 @@ def test_anndata_sparse_compat(tmp_path, diskfmt):
     ad._io.specs.write_elem(f, "/", base)
     adata = ad.AnnData(sparse_dataset(f["/"]))
     assert_equal(adata.X, base)
+
+
+def test_dense_sizeof(ondisk_equivalent_adata, diskfmt):
+    _, _, _, dense_disk = ondisk_equivalent_adata
+
+    size_on_disk = np.array(dense_disk.X.shape).prod() * dense_disk.X.dtype.itemsize
+
+    size_nested_objects = 0
+    for x in ("_obs", "_var"):
+        size_nested_objects += getattr(dense_disk, x).__sizeof__()
+    for x in ("_uns", "_obsm", "_varm", "varp", "_obsp", "_layers"):
+        size_nested_objects += sum(
+            getattr(dense_disk, x)[k].__sizeof__()
+            for k in getattr(dense_disk, x).keys()
+        )
+
+    dense_with_disk = dense_disk.__sizeof__(with_disk=True)
+    dense_without_disk = dense_disk.__sizeof__(with_disk=False)
+
+    assert (
+        dense_with_disk - 128 <= size_on_disk + size_nested_objects <= dense_with_disk
+    )
+    if diskfmt == "h5ad":
+        assert dense_without_disk - 128 <= size_nested_objects <= dense_without_disk
+    else:
+        dense_with_disk == dense_without_disk
```

Comment:
It should be like this, why isn’t that the case?

```suggestion
    assert dense_without_disk - 128 <= size_nested_objects <= dense_without_disk
```

Reply:
It is not the case because when the backend is `h5ad` -> `type(dense_disk.X)` -> `<class 'h5py._hl.dataset.Dataset'>` -> returns the sparse matrix size.
However, when backed is `zarr` -> `type(dense_disk.X)` -> `<class 'numpy.ndarray'>` -> returns the np.array full size.


Testing this, I've also noted a weird thing with `test_dense_sizeof` (with a breakpoint after `dense_without_disk` computation)

```python
diskfmt
'zarr'

dense_disk.__sizeof__(with_disk=True)
30280


 
```

Reply:
It is not the case because when the backend is `h5ad` -> `type(dense_disk.X)` -> `<class 'h5py._hl.dataset.Dataset'>` -> returns the sparse matrix size.
However, when backed is `zarr` -> `type(dense_disk.X)` -> `<class 'numpy.ndarray'>` -> returns the np.array full size.


Testing this, I've also noted a weird thing with `test_dense_sizeof` (with a breakpoint after `dense_without_disk` computation)

```python
>>> diskfmt
'h5py'
>>> dense_with_disk
30152
>>> size_on_disk + size_nested_objects
30088
>>> 152-88
64

>>> diskfmt
'zarr'
>>> dense_disk.__sizeof__(with_disk=True, show_stratified=True)
Size of X      : 20.1kB
Size of .obs   : 5.08kB
Size of .var   : 5.08kB
30280
>>> dense_disk.X.__sizeof__() + dense_disk.obs.__sizeof__() + dense_disk.var.__sizeof__()
30216
>>> size_on_disk + size_nested_objects
30088
>>> for x in ("_uns", "_obsm", "_varm", "varp", "_obsp", "_layers"):
            print(sum(getattr(dense_disk, x)[k].__sizeof__() for k in getattr(dense_disk, x).keys()))
0
0
0
0
0
0
>>> for x in ("X", "_obs", "_var"):
           print(getattr(dense_disk, x).__sizeof__())
20128
5044
5044

>>> np.array(dense_disk.X.shape).prod() * dense_disk.X.dtype.itemsize
20000
>>> dense_disk.X.__sizeof__()
20128
```

We end up with `192B` of difference between the two computation. `128B` of that is explained by np.array class being slightly bigger than its array. However, I don't understand where the extra `64B` comes from as I'm seemingly doing the same computation in the test vs. in the function.

Reply:
Ah, that was my bad, I changed the fallback case to `sys.getsizeof()`, which adds Python’s garbage collector overhead. So it would be more accurate but breaks your tests.